code/datums/components/plumbing/_plumbing.dm

```diff
@@ -0,0 +1,215 @@
+	var/turn_connects = TRUE
+
+/datum/component/plumbing/Initialize(start=TRUE, _turn_connects=TRUE) //turn_connects for wheter or not we spin with the object to change our pipes
+	if(parent && !ismovableatom(parent))
+		return COMPONENT_INCOMPATIBLE
+	var/atom/movable/AM = parent
+	if(!AM.reagents)
+		return COMPONENT_INCOMPATIBLE
+	reagents = AM.reagents
+	turn_connects = _turn_connects
+
+	RegisterSignal(parent, list(COMSIG_MOVABLE_MOVED,COMSIG_PARENT_PREQDELETED), .proc/disable)
+	RegisterSignal(parent, list(COMSIG_OBJ_DEFAULT_UNFASTEN_WRENCH), .proc/toggle_active)
+
+	if(start)
+		enable()
+
+	if(use_overlays)
+		create_overlays()
+
+/datum/component/plumbing/process()
+	if(!demand_connects || !reagents)
+		STOP_PROCESSING(SSfluids, src)
+		return
+	if(reagents.total_volume < reagents.maximum_volume)
+		for(var/D in GLOB.cardinals)
+			if(D & demand_connects)
+				send_request(D)
+///Can we be added to the ductnet?
+/datum/component/plumbing/proc/can_add(datum/ductnet/D, dir)
+	if(!active)
+		return
+	if(!dir || !D)
+		return FALSE
+	if(num2text(dir) in ducts)
+		return FALSE
+
+	return TRUE
+///called from in process(). only calls process_request(), but can be overwritten for children with special behaviour
+/datum/component/plumbing/proc/send_request(dir)
+	process_request(amount = MACHINE_REAGENT_TRANSFER, reagent = null, dir = dir)
+///check who can give us what we want, and how many each of them will give us
+/datum/component/plumbing/proc/process_request(amount, reagent, dir)
+	var/list/valid_suppliers = list()
+	var/datum/ductnet/net
+	if(!ducts.Find(num2text(dir)))
+		return
+	net = ducts[num2text(dir)]
+	for(var/A in net.suppliers)
+		var/datum/component/plumbing/supplier = A
+		if(supplier.can_give(amount, reagent, net))
+			valid_suppliers += supplier
+	for(var/A in valid_suppliers)
+		var/datum/component/plumbing/give = A
+		give.transfer_to(src, amount / valid_suppliers.len, reagent, net)
+///returns TRUE when they can give the specified amount and reagent. called by process request
+/datum/component/plumbing/proc/can_give(amount, reagent, datum/ductnet/net)
+	if(amount <= 0)
+		return
+
+	if(reagent) //only asked for one type of reagent
+		for(var/A in reagents.reagent_list)
+			var/datum/reagent/R = A
+			if(R.type == reagent)
+				return TRUE
+	else if(reagents.total_volume > 0) //take whatever
+		return TRUE
+///this is where the reagent is actually transferred and is thus the finish point of our process()
+/datum/component/plumbing/proc/transfer_to(datum/component/plumbing/target, amount, reagent, datum/ductnet/net)
+	if(!reagents || !target || !target.reagents)
+		return FALSE
+	if(reagent)
+		reagents.trans_id_to(target.parent, reagent, amount)
+	else
+		reagents.trans_to(target.parent, amount, round_robin = TRUE)//we deal with alot of precise calculations so we round_robin=TRUE. Otherwise we get floating point errors, 1 != 1 and 2.5 + 2.5 = 6
```

Comment:
I would suggest scrubbing round_robin stuff, I've edited the code so we're okay with decimals, or at least, aim to.

Reply:
Though, I can understand leaving it TRUE for testing

Reply:
Actually, we don't have round_robin, so this will runtime.